src/grt/grt-wave_options-parse.adb

```diff
@@ -0,0 +1,318 @@
+--  GHDL Run Time (GRT) -  command line options.
+--  Copyright (C) 2002 - 2014 Tristan Gingold
+--
+--  GHDL is free software; you can redistribute it and/or modify it under
+--  the terms of the GNU General Public License as published by the Free
+--  Software Foundation; either version 2, or (at your option) any later
+--  version.
+--
+--  GHDL is distributed in the hope that it will be useful, but WITHOUT ANY
+--  WARRANTY; without even the implied warranty of MERCHANTABILITY or
+--  FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+--  for more details.
+--
+--  You should have received a copy of the GNU General Public License
+--  along with GCC; see the file COPYING.  If not, write to the Free
+--  Software Foundation, 59 Temple Place - Suite 330, Boston, MA
+--  02111-1307, USA.
+--
+--  As a special exception, if other files instantiate generics from this
+--  unit, or you link this unit with other files to produce an executable,
+--  this unit does not by itself cause the resulting executable to be
+--  covered by the GNU General Public License. This exception does not
+--  however invalidate any other reasons why the executable file might be
+--  covered by the GNU Public License.
+
+with System; use System;
+with Grt.Types; use Grt.Types;
+with Grt.Strings; use Grt.Strings;
+with Grt.Vstrings; use Grt.Vstrings;
+with Grt.Errors; use Grt.Errors;
+
+--~ with Grt.Wave_Options.Parse.Debug;
+
+package body Grt.Wave_Options.Parse is
+
+   procedure Start (Option_File : String) is
+      Stream : constant FILEs := File_Open (Option_File);
+      First, Last : Integer;
+      Line : String (1 .. Buf_Size);
+      Lineno : Natural;
+   begin
+      File_Path := new String'(Option_File);
+      Lineno := 0;
+
+      loop -- Reads line after line
+         exit when fgets (Line'Address, Line'Length, Stream) = Null_Address;
+         Lineno := Lineno + 1;
+
+         Last := New_Line_Pos (Line) - 1;
+         if Last < 0 then
+            Last := Line'Last;
+         end if;
+
+         First := First_Non_Whitespace_Pos (Line (Line'First .. Last));
+         if First = -1 or else Line (First) = '#' then
+            goto Continue;
+         end if;
+
+         Last := Last_Non_Whitespace_Pos (Line (First .. Last));
+         declare
+            Line_Str : constant String (1 .. Last + 1 - First) :=
+                       Line (First .. Last);
+         begin
+            Line_Context :=
+               new Line_Context_Type'(Str => new String'(Line_Str),
```

Comment:
On 23/07/16 10:33, Jonsba wrote:
> In src/grt/grt-wave_options-parse.adb
> <https://github.com/tgingold/ghdl/pull/121#discussion_r71969510>:
>
>> +         if Last < 0 then
>> +            Last := Line'Last;
>> +         end if;
>> +
>> +         First := First_Non_Whitespace_Pos (Line (Line'First .. Last));
>> +         if First = -1 or else Line (First) = '#' then
>> +            goto Continue;
>> +         end if;
>> +
>> +         Last := Last_Non_Whitespace_Pos (Line (First .. Last));
>> +         declare
>> +            Line_Str : constant String (1 .. Last + 1 - First) :=
>> +                       Line (First .. Last);
>> +         begin
>> +            Line_Context :=
>> +               new Line_Context_Type'(Str => new String'(Line_Str),
>
> And Line_Context also needs to be an access because if we have :
> /top/sub/a then the tree will looks like : top -> sub -> a. Now if
> Line_Context isn't an access, then in the tree "top", "sub" and "a" will
> have their own copy of Line_Context. I will use some pseudo code to
> explain the problem with that. Here is what we get after the line is
> parsed : top.Line_Context.Max_Level = 1 sub.Line_Context.Max_Level = 2
> a.Line_Context.Max_Level = 3 Where Max_Level is the number of tree
> levels of the line (3 in our case). So we had expected instead :
> top.Line_Context.Max_Level = 3 sub.Line_Context.Max_Level = 3
> a.Line_Context.Max_Level = 3 Which can be only reached when "top", "sub"
> and "a" share a common Line_Context. Actually Line_Context could be
> deallocated right after Parse_Version (Line_Context.Str), or maybe I
> should rather rewrite my code so that Parse_Version doesn't use
> Line_Context anymore but passes the needed informations in a different
> way in the calls to Error_Context and Print_Context.

I am not sure why you really need the Str in Line_Context.
You are saving all significant lines of the file, while I think the line 
number would be enough for error message.

OTOH, the option file shouldn't be very large.

So keep Str.